File: 算法竞赛进阶指南/GoDS (Go Data Structures)/src/math/convolution/setfunction/xorConvolution/xorConvolution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const MOD int = 998244353

func main() {
	// https://judge.yosupo.jp/problem/bitwise_xor_convolution
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n int
	fmt.Fscan(in, &n)
	nums1 := make([]int, 1<<n)
	nums2 := make([]int, 1<<n)
	for i := 0; i < 1<<n; i++ {
		fmt.Fscan(in, &nums1[i])
	}
	for i := 0; i < 1<<n; i++ {
		fmt.Fscan(in, &nums2[i])
	}
	conv := XorConvolution(nums1, nums2)
	for i := 0; i < 1<<n; i++ {
		fmt.Fprint(out, conv[i], " ")
	}
}

func XorConvolution(a, b []int) []int {
	a = append(a[:0:0], a...)
	b = append(b[:0:0], b...)
	a = walshHadamardTransform(a, 1)
	b = walshHadamardTransform(b, 1)
	for i := range a {
		a[i] = a[i] * b[i] % MOD
	}
	res := walshHadamardTransform(a, (MOD+1)/2)
	return res
}

func walshHadamardTransform(f []int, op int) []int {
	n := len(f)
	for l, k := 2, 1; l <= n; l, k = l<<1, k<<1 {
		for i := 0; i < n; i += l {
			for j := 0; j < k; j++ {
				f[i+j], f[i+j+k] = (f[i+j]+f[i+j+k])*op%MOD, (f[i+j]+MOD-f[i+j+k])*op%MOD
			}
		}
	}
	return f
}

File: 19_数学/容斥原理/zeta与mobius/集合zeta/convolution/xorConvolution/xorConvolution.py
from typing import List

from typing import List


MOD = int(1e9) + 7
INV = pow(2, MOD - 2, MOD)


def xorConvolution(a: List[int], b: List[int]) -> List[int]:
    a, b = a[:], b[:]
    a = _walshHadamardTransform(a, 1)
    b = _walshHadamardTransform(b, 1)
    for i in range(len(a)):
        a[i] = a[i] * b[i] % MOD
    res = _walshHadamardTransform(a, INV)
    return res


def _walshHadamardTransform(f, op):
    n = len(f)
    l_, k = 2, 1
    while l_ <= n:
        for i in range(0, n, l_):
            for j in range(k):
                f[i + j], f[i + j + k] = (f[i + j] + f[i + j + k]) * op % MOD, (
                    f[i + j] + MOD - f[i + j + k]
                ) * op % MOD
        l_, k = l_ << 1, k << 1
    return f


if __name__ == "__main__":
    # https://judge.yosupo.jp/submission/129301
    MOD = 998244353
    INV = (MOD + 1) // 2
    n = int(input())
    nums1 = list(map(int, input().split()))
    nums2 = list(map(int, input().split()))
    conv = xorConvolution(nums1, nums2)
    print(*conv)
